### DISCUSSION TITLE
Mounting Disks on Login

### DISCUSSION THREAD

**Discussion by honear** [dokan-dev/dokany#1005] [Q&A]
Hi! Super new here, actually first 'discussion' on Github. I'm trying Dokan mirror and it's doing exactly what I needed, but when I close the command prompt, the drives also get unmounted. Is this the expected behavior? If so, what would I need to do in order to setup up something that mounts the drives on Login and they stay up without the command prompt window always there?

Thank you lots!

**Comment by Liryna** [comment]
Hi @honear,
You can register the mirror as a service using NSSM explained in the second answer here and configure the service to start at login with your user account.
https://stackoverflow.com/questions/3582108/create-windows-service-from-executable

**Reply by honear** [reply]
Oh awesome! Just tried to figure out the final steps. How would I set up the mirror.exe so that it mounts two or more drives with NSSM? Where would I put the "/r \\server\path /l m" stuff?

**Reply by Liryna** [reply]
I never used NSSM but looks like there is a place to put the arguments like "/r \server\path /l m".
If you want multiple mounts, you will need multiple service registered. One for each of them.

**Reply by Agamenox** [reply]
I connect my Buffalo NAS with the mirror command: `mirror.exe /r \\192.168.0.188 /l m` in my lan and add .bat to my task scheduler and  work PERFECTLY in Windows 10!

**Reply by joeu1998** [reply]
@honear I don't suppose you'd be able to share your config for this? I've been tearing my hair out over this all morning. For context: I have a network share which is mapped to B:\ via the Dokan mirror function, using the following syntax:

`mirror.exe /r \\JOECLOUD\Software /l b /v Software`
_So I am mounting the 'Software' share on my server to B:\ on the PC, with the name 'Software'. This all works as expected._

However, configuring this as a service to be automated at login through NSSM is proving to be a headache. I've created a few services now with some different syntax in the 'Arguments' field, but none of them are able to successfully map the share in the same way as when doing it manually via CMD as above. In the NSSM GUI, I've got the following fields:

Application Path: `C:\Program Files\Dokan\Dokan Library-2.0.3\sample\mirror\mirror.exe`
Startup Directory: `C:\Program Files\Dokan\Dokan Library-2.0.3\sample\mirror`
Arguments: `/r \\JOECLOUD\Software /l b /v Software` [I've tried this with `mirror.exe / r...` , `[space]/r...` , `/r...` but nothing seems to work.

The startup type is set to auto, log on account is set to 'Local System account' (I get log on errors when setting this to my own user or administrator accounts)

NSSM then confirms a successful installation. But, when I then go to services.msc and manually click 'start' on the service created, it 'kind of' mounts B:\ , but it is just called 'Local Disk' and when trying to actually open B:\ there is a Windows error of "B:\ is not accessible. The filename, directory name, or volume label syntax is incorrect."

Any advice on where I might be going wrong here would be highly appreciated - Ideally I really need to get these mounts automated on every login / reboot of the VM.

**Reply by joeu1998** [reply]
In addition to the above, for what it's worth, the same behaviour is experienced when pointing NSSM to a .bat file. 

_This .bat file DOES successfully mount the share at B:\ when executed as usual and CMD is left open._

However, when setting up this .bat file as a service in NSSM, the same behaviour happens of B:\ is kinda half mounted, but not actually accessible. (I have tested with other shares, as well - same results).

**Reply by The009** [reply]
How To Using NSSM
Download Latest Version:  [nssm 2.24-101-g897c7ad](https://nssm.cc/ci/nssm-2.24-101-g897c7ad.zip) (2017-04-26) (As of this moment)
- Extract Zip
- Launch CMD Window
- Go To NSSM
`CD C:\Users\the009\Downloads\nssm-2.24-101-g897c7ad\win64`

- Launch the Installer with the Service Name
`nssm.exe install MountA`
- Allow it to run as administrator ( Required to add service )

### GUI Time
**Application Tab**
- Path : `C:\Program Files\Dokan\Dokan Library-2.0.6\sample\mirror\mirror.exe`
- Startup Directory:  `C:\Program Files\Dokan\Dokan Library-2.0.6\sample\mirror`
- Arguments "" Are Needed : `/v "Video" /r "\\10.0.2.100\Videos" /l A`

**Details Tab**
- Display Name : `Mount Drive A`
- Description : `Mount SAMBA Share To Local Drive`
- Service Type: `Automatic (Delayed Start)`
  _This gives Windows time to start the samba client and network services before trying to mount the drives_

**Process Tab**
- Uncheck Console Window Check Box

** DONE **
- Press Install Service

(You can now reboot to test it or start the service manually to test it )

**Comment by Dwedit** [comment]
Note that if you want a Read-Write mirror, you can do this without a user-mode file system, NTFS provides this for free.  You can create a symbolic link or junction point, and it's easy to do with tools like Link Shell Extension.

NTFS's built-in features won't give you a read only mirror though.

**Comment by Kubas556** [comment]
Any update on this? I was able to create service which starts at startup, but when i try to open the mounted disk, its not accessible.

**Reply by The009** [reply]
Make sure you typed the mount location correctly.
The service will be created even if you don't have a proper connection.

**Reply by Kubas556** [reply]
I actually have custom project based on the mirror example. I use API calls to my server, which returns files/folder structure instead of mirror physical disk. Maybe there is some network access problem? I run the service with network service account, but that did not help. Windows event viewer shows no error after service startup

**Reply by Olof-Lagerkvist** [reply]
If the mounted file system shows up but is not accessible, I would suggest that you add some logging to your implementation and try to find out whether it fails in some way when files or directories are requested. It is quite common to get permission issues of various kinds when running as services compared to running as an interactive user.

**Reply by lane-ftw** [reply]
> Any update on this? I was able to create service which starts at startup, but when i try to open the mounted disk, its not accessible.

I think it's a permissions/ login issue. I think it's trying to sign in as SYSTEM (as that's the user running the service on windows) and I'm not sure what password it's trying to use or how to set it. (Assuming you've followed @The009's instructions and used nssm.)
If you're using samba, in your smb.conf you can add `guest ok = yes` to get it to work. Note that this is a security risk and you should set up something like `valid users = <USERNAME>` or `valid users = @<GROUP>`. 

EDIT: According to this thread https://github.com/dokan-dev/dokany/discussions/1203 there is no way to set a user/password combo when connecting to a share. So running as a service seems like kind of a bad idea with no other security implemented. Perhaps making a share that only allows SYSTEM to login from a specific IP would be a way to do _something_ other than leaving it completely open. `allow hosts = xxx.xxx.xxx.xxx` should do that.

**Reply by Kubas556** [reply]
Thanks for answers, i just created a .bat file and put it inside windows startup folder XD
It's the easiest way.

**Reply by The009** [reply]
> EDIT: According to this thread #1203 there is no way to set a user/password combo when connecting to a share. So running as a service seems like kind of a bad idea with no other security implemented. Perhaps making a share that only allows SYSTEM to login from a specific IP would be a way to do _something_ other than leaving it completely open. `allow hosts = xxx.xxx.xxx.xxx` should do that.

^^ If you have set the Login information in Windows you can use shares that are NOT open to everyone, The problem with Windows is that you are limited to ONE single login for a network. So make sure whatever user you create has access to everything that PC needs access to, then set the credentials in Windows. Once the network services start (Hence the delayed start) the login information is then used by Windows to access that recourse.

**Comment by Olof-Lagerkvist** [comment]
> EDIT: According to this thread #1203 there is no way to set a user/password combo when connecting to a share. So running as a service seems like kind of a bad idea with no other security implemented. Perhaps making a share that only allows SYSTEM to login from a specific IP would be a way to do _something_ other than leaving it completely open. `allow hosts = xxx.xxx.xxx.xxx` should do that.

It is still possible to run as a service, but you need to configure the service to run under a user account that is also present with the same password at the computer where the network resource is shared from. Running the service under SYSTEM account and accessing network shares is not at all a trivial task, though.

**Comment by gabrielxfx** [comment]
just on NMSS add use your local computer account to start the service ....

say

gabriel
1234      1234

and that's it
